doc/tutorials/setup_openstack_libvirt.rst

```diff
@@ -12,25 +12,23 @@ to avoid crashing your system and to take advantage of easy
 snapshots/restores so that you can quickly roll back your system in the
 event of setup failure. (You should only install OpenStack directly on Ubuntu if
 you have a dedicated testing machine.) This setup utilizes LXC/LXD on
-Ubuntu 16.04 or 18.04.
+Ubuntu 18.04.
 
 Install ACRN
 ************
 
-#. Install ACRN using Ubuntu 16.04 or 18.04 as its Service VM.
-
-   .. important:: Need instructions from deleted document (using Ubuntu
-      as SOS)
+#. Install ACRN using Ubuntu 18.04 as its Service VM. Refer to
+   :ref:`Build and Install ACRN on Ubuntu <build-and-install-acrn-on-ubuntu>`.
 
 #. Make the acrn-kernel using the `kernel_config_uefi_sos
    <https://raw.githubusercontent.com/projectacrn/acrn-kernel/master/kernel_config_uefi_sos>`_
    configuration file (from the ``acrn-kernel`` repo).
 
-#. Add the following kernel boot arg to give the Service VM  more loop
-   devices. Refer to `Kernel Boot Parameters
+#. Add the following kernel boot arg to give the Service VM  more memory
+   and more loop devices. Refer to `Kernel Boot Parameters
    <https://wiki.ubuntu.com/Kernel/KernelBootParameters>`_ documentation::
 
-      max_loop=16
+      hugepagesz=1G hugepages=10 max_loop=16
```

Comment:
That does not work on my system, the parameters added to `GRUB_CMDLINE_LINUX_DEFAULT` in `/etc/default/grub` are not taken into account. I believe one _could_ add those via the Grub command-line interface (never tried) but this would not be permanent.

I believe the right way for ACRN would actually be to change the `bootargs` parameter in the board/scenario configuration files.
From https://projectacrn.github.io/latest/introduction/index.html#boot-sequence: *In this boot mode, the boot options of pre-launched VM and service VM are defined in the variable of bootargs of struct vm_configs[vm id].os_config in the source code misc/vm_configs/$(SCENARIO)/vm_configurations.c by default. Their boot options can be overridden by the GRUB menu*

Having said that, ACRN also already integrates some logic to automatically reserve hugepages - see https://github.com/projectacrn/acrn-hypervisor/blob/802065cf2fcde4e623a07e8f0a46cb8f7a5eec6f/hypervisor/common/vm_load.c#L147-L160. Do we really need to override that logic?

If not, we're left with the `max_loop` parameter that we need to add somewhere.

Reply:
@gvancuts the default hugepagesz is 12 because NUM_REMAIN_1G_PAGES=3 in acrn-hypervisor/hypervisor/common/vm_load.c. This causes there is not enough memory to setup openstack in SOS when I try to setup openstack with ACRN 2.3(OOM issue). So we need to decrease the size of hugepagesz. 
updating GRUB_CMDLINE_LINUX_DEFAULT in /etc/default/grub would not be taken into account.
There are 2 ways to decrease the size of hugepagesz: 
1) set NUM_REMAIN_1G_PAGES>3 in acrn-hypervisor/hypervisor/common/vm_load.c, where we need an offline patch to achieve this.
2) we can overwrite the size of hugepagesz from setting grub config in /etc/grub.d/40_custom like this:
`menuentry "ACRN Multiboot Ubuntu Service VM" --id ubuntu-service-vm {
  recordfail
  load_video
  insmod gzio
  insmod part_gpt
  insmod ext2

  echo 'loading ACRN...'
  multiboot2 /boot/acrn/acrn.bin
  module2 /boot/bzImage Linux_bzImage root=/dev/sda3 console=ttyS0 idle=halt rw rootwait console=tty0 consoleblank=0 no_timer_check quiet loglevel=3 i915.nuclear_pageflip=1 hvlog=2M@0xe00000 memmap=0x200000\$0xe00000 maxcpus=4 **max_loop=16 hugepagesz=1G hugepages=10**
}
`
then update-grub which I have validated that it will take effect to increase the available memory for SOS.
ref steps: https://projectacrn.github.io/latest/getting-started/rt_industry_ubuntu.html#update-grub-for-the-ubuntu-service-vm

I choose the second way in the doc and put the args "hugepagesz=1G hugepages=10 max_loop=16" into module2 part of grub config which is the minimal change for the setup guide.

Reply:
Can you explain these steps in this document? The link to the Ubuntu article does not talk about these.